Issue #10956 - Reviewed Expect 100 Continue Handler

gregw opened the pull request:
Removed duplicate handling from servlet. Now only handled in core on first demand. Added additional tests.

gregw committed:
Reviewed Expect 100 Continue Handler for #10956

Removed duplicate handling from servlet. Now only handled in core on first demand.
Added additional tests.

gregw commented:
@sbordet I'm not sure we have 100-Continue handling correct for h2.  See https://httpwg.org/specs/rfc9110.html#field.expect
I think we are good when we do send a 100-Continue response, but I'm not so sure when we don't (i.e. we send a response without reading the request).

With HTTP/1, we handle that by forcing the connection to be closed as soon as the response is sent and not waiting for the request content.

But with h2, there is no equivalent tear down and ignore the request content.      Maybe we should either have to send the 100-continues response OR reset the stream??? But then how could we send a redirect to such a POST?

sbordet commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/HttpClientContinueTest.java:
```diff
@@ -113,17 +114,83 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
+                // read/write first byte to send 100-Continue
                 ServletInputStream input = request.getInputStream();
-                // Make sure we chunk the response too
+                response.getOutputStream().write(input.read());
```
I don't like that now we have to read() in order to send the `100` interim response.

sbordet reviewed (changes requested):
@gregw I think the code duplication is necessary to keep the semantic that calling `getInputStream()` sends the `100` interim response.

Having to call `read()` is probably breaking user's code (like it breaks our tests if we don't add a `read()`).

We do only send the `100` interim response once, so I don't see the problem in that mechanism.

We could perhaps move the code a static helper, e.g. `Request.process100(request)` to reduce duplication.

I'm ok adding RST_STREAM for HTTP/2 and HTTP/3 after the response, but only in cases where the application really does not read the request content.
To be clear the case of:

* write 100 interim (explicitly in core, or via `getInputStream()`)
* do not ready yet, but write the response
* read the request content

is a legit one and we should not send the RST_STREAM.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10956/Expect100Continue

gregw committed:
Updates from review

added the willRead method to HttpStream and Request to allow signal that a read will be attempted.

gregw commented:
@sbordet I had an attempt at making the H2 stream send a reset(NO_ERROR) (immediately if locally closed or later when a close happens).   But I fail a lot of tests as the client treats the reset as an abort, calling executeAbort.     We need to modify the client  so that a reset(NO_ERROR) is not seen as a failure.  Do we have a way of telling a content source "thanks, but no more content is needed and this is not an error"?

Do you want to work on this?   If so, do you want to do it in this PR or another one. (edit moved to #11029)

gregw commented:
@sbordet @lorban I will do the reset(NO_ERROR) work in a different PR.  But it needs to build on this, so can you review this as is?

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10956/Expect100Continue

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10956/Expect100Continue

gregw commented:
@sbordet @lorban nudge (CI failures are unrelated)

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1331,6 +1331,16 @@ public String getId()
+            if (_expects100Continue)
+            {
+                _expects100Continue = false;
+                send(_request, HttpGenerator.CONTINUE_100_INFO, false, null, Callback.NOOP);
```
Now this makes me wonder if this isn't racing with a user-generated write? I mean such handler:

```java
boolean handle(Request req, Response resp, Callback callback)
{
  req.demand(() -> {}); // this triggers a write with a NOOP callback
  resp.write(true, BufferUtil.EMPTY, callback); // doesn't this risk a WritePendingException?
}
```

gregw replied:
Good question, will check.

gregw replied:
Currently we protect against simultaneous writes by looking at the `_writeCallback` field in the `ChannelResponse` with a lock.  If it is non-null, then we throw a write pending exception.

This doesn't work for a `writeInterim` for a number of reason, firstly it for some reason has a different style of API (with a `CompletableFuture`), which makes it complicated to be part of the same mechanism.  But more importantly, I do not think that it is a WPE if a write and a read are done at the same time.   So we would need different protection anyway.

Currently sending an interim response ends up being HttpStream send (with noop callback) that could be concurrent with a send from a write, so it is up to the stream impl how that is handled.   HTTP/1 at least looks like it doesn't and there could be issues with 2 threads in the generator.

@sbordet @lorban need to hangout on this one.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpStream.java:
```diff
@@ -46,6 +46,22 @@ public interface HttpStream extends Callback
+     *     state to not expect the request content to be sent and to ignore any further content sent on the channel</li>
+     * </ul>
+     */
+    default void willRead()
```
I think this shouldn't be marked as `default`.

lorban commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -254,6 +254,14 @@ static long getTimeStamp(Request request)
+     * Signal the intention to read the request content for the purposes of 100-Continues processing.
+     * @see HttpStream#willRead()
+     */
+    default void willRead()
```
I think this shouldn't be marked as `default`.

lorban commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/HttpClientContinueTest.java:
```diff
@@ -111,6 +112,68 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
+        testExpect100ContinueRespond100ContinueBlocking(transport, "data1".getBytes(StandardCharsets.UTF_8), "data2".getBytes(StandardCharsets.UTF_8), "data3".getBytes(StandardCharsets.UTF_8));
+    }
+
+    private void testExpect100ContinueRespond100ContinueBlocking(Transport transport, byte[]... contents) throws Exception
```
This private method looks like it could be inlined in the test method.

lorban commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/HttpClientContinueTest.java:
```diff
@@ -111,6 +112,68 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
+
+    private void testExpect100ContinueRespond100ContinueBlocking(Transport transport, byte[]... contents) throws Exception
+    {
+        start(transport, new HttpServlet()
```
If you store the reader thread like this:

```java
AtomicReference<Thread> readerThreadRef = new AtomicReference<>();
start(transport, new HttpServlet()
{
    @Override
    protected void service(HttpServletRequest request, HttpServletResponse response) throws IOException
    {
        readerThreadRef.set(Thread.currentThread());
        // Send 100-Continue and copy the content back
        IO.copy(request.getInputStream(), response.getOutputStream());
    }
});
```

then...

lorban commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/HttpClientContinueTest.java:
```diff
@@ -111,6 +112,68 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
+                    try
+                    {
+                        // Sleep here to ensure that sometimes the reader will block/pause even after sending 100.
+                        Thread.sleep(100);
```
...that `sleep` can be replaced with:

```java
await().atMost(5, TimeUnit.SECONDS).until(() ->
{
    Thread thread = readerThreadRef.get();
    if (thread == null)
        return false;
    return thread.getState() == Thread.State.WAITING;
});
```

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -254,6 +254,14 @@ static long getTimeStamp(Request request)
+     * Signal the intention to read the request content for the purposes of 100-Continues processing.
+     * @see HttpStream#willRead()
+     */
+    default void willRead()
+    {
+    }
```
I think we can avoid adding this method, as it is only called from EEN, and we can call `writeInterim()` instead (perhaps have a `static write100()` utility method in `Response`).

The state about the 100 is kept in the `HttpStream` where we can easily decide whether to actually send it or not, so we won't need the extra methods in `Request` and `HttpStream`.

gregw replied:
@sbordet  So are you saying we change the semantic of `writeInterim(...)` to `sendInterimIfAppropriate(...)` ?   I .e. it only sends the interim after checking if it OK to send it.

I'll give it a go, but is it really, better than `willRead`?   Without a rename, it will be very confusing.

sbordet replied:
No I'm saying to call `writeInterim()` as it is.

It will call `HttpStream.send()`.

There, we have all the state to figure out whether the interim is 100, whether we had the `Expect` header for 100, etc. and if the interim must really be written (or if it has already).

100 is a special case of interim as you can't send it more than once, it needs an `Expect` header, etc.
103, is very different in that regard and needs no state.

So the "if appropriate" is implicit for 100 because of what it is, and it does not apply to other interims.

It is better than `willRead()` because it's it won't force to add a new method in `Request` and `HttpStream`.

It won't be confusing because we can always fail the callback/completable if the 100 is not written for some reason (e.g. already sent).

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12/10956/Expect100Continue

gregw committed:
updates from review

gregw committed:
updates from review

Moved all 100 continues handling into HttpChannelState and protected it from concurrent writes

gregw committed:
updates from review

Moved all 100 continues handling into HttpChannelState and protected it from concurrent writes

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -883,28 +886,43 @@ public void demand(Runnable demandCallback)
+            }
+            else
+            {
+                stream.send(_metaData, new MetaData.Response(100, null, getConnectionMetaData().getHttpVersion(), HttpFields.EMPTY), false, null, interimCallback);
```
I'd prefer `HttpStatus.CONTINUE_100` instead of just the number.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -883,28 +886,43 @@ public void demand(Runnable demandCallback)
+            else
+            {
+                stream.send(_metaData, new MetaData.Response(100, null, getConnectionMetaData().getHttpVersion(), HttpFields.EMPTY), false, null, interimCallback);
+                interimCallback.thenRun(stream::demand);
```
`thenRun()` is only executed if the CF succeeds. If it fails, should we not demand?

If we should demand in any case, use `whenComplete((r, x) -> stream.demand())`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1017,6 +1035,8 @@ public String toString()
      */
     public static class ChannelResponse implements Response, Callback
     {
+        public static final CompletableFuture<Void> UNEXPECTED_100_CONTINUE = CompletableFuture.failedFuture(new IllegalStateException("100 not expected"));
+        public static final CompletableFuture<Void> COMMITTED_100_CONTINUE = CompletableFuture.failedFuture(new IllegalStateException("Committed"));
```
Make them `private`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1117,25 +1137,31 @@ public void write(boolean last, ByteBuffer content, Callback callback)
+                        if (_writeCallback instanceof InterimCallback interimCallback)
+                        {
+                            // Do this write after the interim callback.
+                            interimCallback.thenRun(() -> write(last, content, callback));
```
Again, `thenRun()` only runs if CF succeeds. If fails don't we have to at least fail the write callback?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1288,19 +1314,36 @@ public void reset()
+                HttpChannelState httpChannelState = _request.lockedGetHttpChannelState();
+                stream = httpChannelState._stream;
+
+                if (status == 100)
```
Use `HttpStatus`.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiRequest.java:
```diff
@@ -790,10 +790,8 @@ public ServletInputStream getInputStream() throws IOException
-        if (getServletRequestInfo().getServletChannel().isExpecting100Continue())
-            getServletRequestInfo().getServletChannel().continue100(getServletRequestInfo().getHttpInput().available());
-
+        // write a 100 continue, but ignore the response if it was not expected.
```
```suggestion
        // Try to write a 100 continue, ignoring if it was not necessary.
```

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiRequest.java:
```diff
@@ -1074,7 +1072,12 @@ public String toString()
-        if (_reader == null || !charset.equals(_readerCharset))
+        if (_reader != null && charset.equals(_readerCharset))
+        {
+            // write a 100 continue, but ignore the response if it was not expected.
```
```suggestion
            // Try to write a 100 continue, ignoring if it was not necessary.
```

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiRequest.java:
```diff
@@ -790,10 +790,8 @@ public ServletInputStream getInputStream() throws IOException
-            getServletRequestInfo().getServletChannel().continue100(getServletRequestInfo().getHttpInput().available());
-
+        // write a 100 continue, but ignore the response if it was not expected.
+        _servletChannel.getResponse().writeInterim(100, HttpFields.EMPTY);
```
Use `HttpStatus`.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletApiRequest.java:
```diff
@@ -1074,7 +1072,12 @@ public String toString()
+        if (_reader != null && charset.equals(_readerCharset))
+        {
+            // write a 100 continue, but ignore the response if it was not expected.
+            _servletChannel.getResponse().writeInterim(100, HttpFields.EMPTY);
```
Use `HttpStatus`.

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Request.java:
```diff
@@ -825,10 +825,8 @@ public ServletInputStream getInputStream() throws IOException
-        if (_channel.isExpecting100Continue())
-            _channel.send100Continue(_input.available());
-
+        // write a 100 continue, but ignore the response if it was not expected.
```
```suggestion
        // Try to write a 100 continue, ignoring if it was not necessary.
```

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Request.java:
```diff
@@ -1050,7 +1048,12 @@ public BufferedReader getReader() throws IOException
-        if (_reader == null || !encoding.equalsIgnoreCase(_readerEncoding))
+        if (_reader != null && encoding.equalsIgnoreCase(_readerEncoding))
+        {
+            // write a 100 continue, but ignore the response if it was not expected.
```
```suggestion
            // Try to write a 100 continue, ignoring if it was not necessary.
```

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Request.java:
```diff
@@ -825,10 +825,8 @@ public ServletInputStream getInputStream() throws IOException
-            _channel.send100Continue(_input.available());
-
+        // write a 100 continue, but ignore the response if it was not expected.
+        _channel.getCoreResponse().writeInterim(100, HttpFields.EMPTY);
```
Use `HttpStatus`.

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Request.java:
```diff
@@ -1050,7 +1048,12 @@ public BufferedReader getReader() throws IOException
+        if (_reader != null && encoding.equalsIgnoreCase(_readerEncoding))
+        {
+            // write a 100 continue, but ignore the response if it was not expected.
+            _channel.getCoreResponse().writeInterim(100, HttpFields.EMPTY);
```
Use `HttpStatus`.

lorban reviewed (changes requested):
Same nits as @sbordet

gregw committed:
updates from review

Moved all 100 continues handling into HttpChannelState and protected it from concurrent writes

sbordet committed:
Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
